src/py/pyodide/_package_loader.py

```diff
@@ -16,38 +16,97 @@
+    )
+
+
+def get_format(format: str) -> str:
+    for (fmt, extensions, _) in shutil.get_unpack_formats():
+        if format == fmt:
+            return fmt
+        if format in extensions:
+            return fmt
+        if "." + format in extensions:
+            return fmt
+    raise ValueError(f"Unrecognized format {format}")
+
+
+def unpack_buffer(
+    buffer: JsProxy,
+    *,
+    filename: str = "",
+    format: str = None,
+    target: Literal["site", "lib", None] = None,
+    extract_dir: str = None,
+    calculate_dynlibs: bool = False,
+) -> Optional[JsProxy]:
     """Used to install a package either into sitepackages or into the standard
     library.
 
     This is a helper method called from ``loadPackage``.
 
     Parameters
     ----------
+    buffer
+        A Javascript ``Uint8Array`` with the binary data for the archive.
+
     filename
         The name of the file we are extracting. We only care about it to figure
-        out whether the buffer represents a tar file or a zip file.
+        out whether the buffer represents a tar file or a zip file. Ignored if
+        format argument is present.
 
-    buffer
-        A Javascript ``Uint8Array`` with the binary data for the archive.
+    format
+        Controls the format that we assume the archive has. Overrides the file
+        extension of filename.
+
+    extract_dir
+        Controls which directory the file is unpacked into. Default is the
+        working directory. Mutually exclusive with target.
 
     target
-        Controls which directory the package is installed into. Either "site"
-        which installs the package into the sitepackages directory or "lib"
-        which installs the package into the standard library.
+        Controls which directory the file is unpacked into. Either "site" which
+        unpacked the file into the sitepackages directory or "lib" which
+        unpacked the file into the standard library. Mutually exclusive with
+        extract_dir.
+
+    calculate_dynlibs
+        If true, will return a Javascript Array of paths to dynamic libraries
+        ('.so' files) that were in the archive. We need to precompile these Wasm
+        binaries in `load-pyodide.js`. These paths point to the unpacked
+        locations of the .so files.
 
     Returns
     -------
-        A Javascript Array of paths to dynamic libraries ('.so' files) that were
-        in the archive. We need to precompile these Wasm binaries in
-        `load-pyodide.js`. These paths point to the unpacked locations of the
-        .so files.
+        If calculate_dynlibs is True, a Javascript Array of dynamic libraries.
+        Otherwise, return None.
+
     """
-    target_dir = TARGETS[target]
+    if format:
+        format = get_format(format)
+    if target and extract_dir:
+        raise ValueError("Cannot provide both 'target' and 'extract_dir'")
+    if not filename and not format:
```

Comment:
```suggestion
    if not filename and format is None:
```
technically format="" would be provided.

Reply:
Well if someone says `format=""` we will throw an error further down anyways, but I guess it's an invalid format error rather than "you didn't give us a format"